File: interop-testing/src/test/java/io/grpc/testing/integration/ProxyTest.java
/*
 * Copyright 2016 The gRPC Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.grpc.testing.integration;

import static org.junit.Assert.assertEquals;

import com.google.common.io.ByteStreams;
import io.netty.util.concurrent.DefaultThreadFactory;
import java.io.DataInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;
import org.junit.After;
import org.junit.AfterClass;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

@RunWith(JUnit4.class)
public class ProxyTest {

  private static ThreadPoolExecutor executor =
      new ThreadPoolExecutor(8, 8, 1, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(),
          new DefaultThreadFactory("proxy-test-pool", true));

  private TrafficControlProxy proxy;
  private Socket client;
  private Server server;

  @AfterClass
  public static void stopExecutor() {
    executor.shutdown();
  }

  @After
  public void shutdownTest() throws IOException {
    proxy.shutDown();
    server.shutDown();
    client.close();
  }

  @Test
  @org.junit.Ignore // flaky. latency commonly too high
  public void smallLatency() throws Exception {
    server = new Server();
    int serverPort = server.init();
    executor.execute(server);

    int latency = (int) TimeUnit.MILLISECONDS.toNanos(50);
    proxy = new TrafficControlProxy(serverPort, 1024 * 1024, latency, TimeUnit.NANOSECONDS);
    proxy.start();
    client = new Socket("localhost", proxy.getPort());
    client.setReuseAddress(true);
    OutputStream clientOut = client.getOutputStream();
    DataInputStream clientIn = new DataInputStream(client.getInputStream());
    byte[] message = new byte[1];

    // warmup
    for (int i = 0; i < 5; i++) {
      clientOut.write(message, 0, 1);
    }
    clientIn.readFully(new byte[5]);

    // test
    List<Long> rtts = new ArrayList<>();
    for (int i = 0; i < 3; i++) {
      long start = System.nanoTime();
      clientOut.write(message, 0, 1);
      clientIn.read(message);
      rtts.add(System.nanoTime() - start);
    }
    Collections.sort(rtts);
    long rtt = rtts.get(0);
    assertEquals(latency, rtt, .5 * latency);
  }

  @Test
  public void bigLatency() throws Exception {
    server = new Server();
    int serverPort = server.init();
    executor.execute(server);

    int latency = (int) TimeUnit.MILLISECONDS.toNanos(250);
    proxy = new TrafficControlProxy(serverPort, 1024 * 1024, latency, TimeUnit.NANOSECONDS);
    proxy.start();
    client = new Socket("localhost", proxy.getPort());
    OutputStream clientOut = client.getOutputStream();
    DataInputStream clientIn = new DataInputStream(client.getInputStream());
    byte[] message = new byte[1];

    // warmup
    for (int i = 0; i < 5; i++) {
      clientOut.write(message, 0, 1);
    }
    clientIn.readFully(new byte[5]);

    // test
    List<Long> rtts = new ArrayList<>();
    for (int i = 0; i < 2; i++) {
      long start = System.nanoTime();
      clientOut.write(message, 0, 1);
      clientIn.read(message);
      rtts.add(System.nanoTime() - start);
    }
    Collections.sort(rtts);
    long rtt = rtts.get(0);
    assertEquals(latency, rtt, .5 * latency);
  }

  @Test
  public void smallBandwidth() throws Exception {
    server = new Server();
    int serverPort = server.init();
    server.setMode("stream");
    executor.execute(server);

    int bandwidth = 64 * 1024;
    proxy = new TrafficControlProxy(serverPort, bandwidth, 200, TimeUnit.MILLISECONDS);
    proxy.start();
    client = new Socket("localhost", proxy.getPort());
    DataInputStream clientIn = new DataInputStream(client.getInputStream());

    clientIn.readFully(new byte[100 * 1024]);
    int sample = bandwidth / 5;
    List<Double> bandwidths = new ArrayList<>();
    for (int i = 0; i < 5; i++) {
      long start = System.nanoTime();
      clientIn.readFully(new byte[sample]);
      long duration = System.nanoTime() - start;
      double actualBandwidth = sample / (((double) duration) / TimeUnit.SECONDS.toNanos(1));
      bandwidths.add(actualBandwidth);
    }
    Collections.sort(bandwidths);
    double bandUsed = bandwidths.get(bandwidths.size() - 1);
    assertEquals(bandwidth, bandUsed, .5 * bandwidth);
  }

  @Test
  public void largeBandwidth() throws Exception {
    server = new Server();
    int serverPort = server.init();
    server.setMode("stream");
    executor.execute(server);
    int bandwidth = 10 * 1024 * 1024;
    proxy = new TrafficControlProxy(serverPort, bandwidth, 200, TimeUnit.MILLISECONDS);
    proxy.start();
    client = new Socket("localhost", proxy.getPort());
    DataInputStream clientIn = new DataInputStream(client.getInputStream());

    clientIn.readFully(new byte[100 * 1024]);
    int sample = bandwidth / 5;
    List<Double> bandwidths = new ArrayList<>();
    for (int i = 0; i < 5; i++) {
      long start = System.nanoTime();
      clientIn.readFully(new byte[sample]);
      long duration = System.nanoTime() - start;
      double actualBandwidth = sample / (((double) duration) / TimeUnit.SECONDS.toNanos(1));
      bandwidths.add(actualBandwidth);
    }
    Collections.sort(bandwidths);
    double bandUsed = bandwidths.get(bandwidths.size() - 1);
    assertEquals(bandwidth, bandUsed, .5 * bandwidth);
  }

  // server with echo and streaming modes
  private static class Server implements Runnable {
    private ServerSocket server;
    private String mode = "echo";

    public void setMode(String mode) {
      this.mode = mode;
    }

    /**
     * Initializes server and returns its listening port.
     */
    public int init() throws IOException {
      server = new ServerSocket(0);
      return server.getLocalPort();
    }

    public void shutDown() throws IOException {
      server.close();
    }

    @Override
    public void run() {
      try {
        Socket rcv = server.accept();
        try {
          handleSocket(rcv);
        } finally {
          rcv.close();
        }
      } catch (IOException e) {
        throw new RuntimeException(e);
      }
    }

    private void handleSocket(Socket rcv) throws IOException {
      InputStream serverIn = rcv.getInputStream();
      OutputStream serverOut = rcv.getOutputStream();
      if (mode.equals("echo")) {
        ByteStreams.copy(serverIn, serverOut);
      } else if (mode.equals("stream")) {
        byte[] message = new byte[1024];
        while (true) {
          try {
            serverOut.write(message);
          } catch (IOException ignored) {
            // Client closed
            break;
          }
        }
      } else {
        throw new RuntimeException("Unknown mode: use 'echo' or 'stream'");
      }
    }
  }
}

File: internal/transport/proxy_test.go
//go:build !race
// +build !race

/*
 *
 * Copyright 2017 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package transport

import (
	"context"
	"net"
	"net/http"
	"net/netip"
	"testing"
	"time"

	"google.golang.org/grpc/internal/proxyattributes"
	"google.golang.org/grpc/internal/testutils"
	"google.golang.org/grpc/internal/testutils/proxyserver"
	"google.golang.org/grpc/resolver"
)

func (s) TestHTTPConnectWithServerHello(t *testing.T) {
	serverMessage := []byte("server-hello")
	blis, err := testutils.LocalTCPListener()
	if err != nil {
		t.Fatalf("failed to listen: %v", err)
	}
	reqCheck := func(req *http.Request) {
		if req.Method != http.MethodConnect {
			t.Errorf("unexpected Method %q, want %q", req.Method, http.MethodConnect)
		}
		host, _, err := net.SplitHostPort(req.URL.Host)
		if err != nil {
			t.Error(err)
		}
		_, err = netip.ParseAddr(host)
		if err != nil {
			t.Error(err)
		}
	}
	pServer := proxyserver.New(t, reqCheck, true)

	msg := []byte{4, 3, 5, 2}
	recvBuf := make([]byte, len(msg))
	done := make(chan error, 1)
	go func() {
		in, err := blis.Accept()
		if err != nil {
			done <- err
			return
		}
		defer in.Close()
		in.Write(serverMessage)
		in.Read(recvBuf)
		done <- nil
	}()

	// Dial to proxy server.
	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
	defer cancel()
	c, err := proxyDial(ctx, resolver.Address{Addr: pServer.Addr}, "test", proxyattributes.Options{ConnectAddr: blis.Addr().String()})
	if err != nil {
		t.Fatalf("HTTP connect Dial failed: %v", err)
	}
	defer c.Close()
	c.SetDeadline(time.Now().Add(defaultTestTimeout))

	// Send msg on the connection.
	c.Write(msg)
	if err := <-done; err != nil {
		t.Fatalf("Failed to accept: %v", err)
	}

	// Check received msg.
	if string(recvBuf) != string(msg) {
		t.Fatalf("Received msg: %v, want %v", recvBuf, msg)
	}

	if len(serverMessage) > 0 {
		gotServerMessage := make([]byte, len(serverMessage))
		if _, err := c.Read(gotServerMessage); err != nil {
			t.Errorf("Got error while reading message from server: %v", err)
			return
		}
		if string(gotServerMessage) != string(serverMessage) {
			t.Errorf("Message from server: %v, want %v", gotServerMessage, serverMessage)
		}
	}
}
